### PR TITLE
feat(lsp): add inlay hint label formatter

### LINKED ISSUE #27240: Truncate long inlay hints
### Problem

https://github.com/microsoft/vscode/issues/199610 explains the problem, but the TLDR is that some language servers (like the TypeScript one) don't perform any truncation on inlay hints, which might result on several lines of virtual text.

### Expected behavior

My proposal is to add a configuration setting (and hence easily customizable with `vim.lsp.with`) that limits the length of inlay hints.

### PR DESCRIPTION
**Pull request by MariaSolOs**
Closes #27240.
(I got a nonzero number of likes in the issue hence I'm assuming this is a reasonable request).

For better flexibility, I decided to expose a function (instead of just a truncation limit) for manipulating the processed inlay hint label.

### CODE REVIEW

**Comment by MariaSolOs** [comment]
Already asked this on Matrix but still putting it here to gather everyone's thoughts: While working on this, I realized that customizing the display of inlay hints is already quite straightforward by overriding the handler, and I really don't want to make the same mistakes from `vim.lsp.util` of providing redundant helpers.

wdy all think about this? Does it make sense to provide commonly-used LSP handler options even if they can be easily (< 10 lines of code) overriden?

**Comment by noahfrederick** [comment]
> wdy all think about this? Does it make sense to provide commonly-used LSP handler options even if they can be easily (< 10 lines of code) overriden?

The underlying motivation for adding this setting tells me that the default behavior should be improved instead, at least as a first step. Admittedly, I don't have a full grasp of the tradeoffs here, though. It seems like we would want a mechanism in place to reveal the full inlay hint before going ahead and truncating by default, for example.

**Comment by MariaSolOs** [comment]
> The underlying motivation for adding this setting tells me that the default behavior should be improved instead, at least as a first step. Admittedly, I don't have a full grasp of the tradeoffs here, though. It seems like we would want a mechanism in place to reveal the full inlay hint before going ahead and truncating by default, for example.

The default behavior is to just display the label as is (no truncation).

I do agree that the editor doesn't have enough semantic information to shorten the label in a "smart" way, and that each language server should be the one that modifies the label. However this would result in duplicated settings and implementations for the same thing, which isn't great either.

File: runtime/lua/vim/lsp/inlay_hint.lua [hunk 1]
```diff
@@ -4,11 +4,17 @@ local ms = require('vim.lsp.protocol').Methods
 local api = vim.api
 local M = {}
 
+---@class vim.lsp.inlay_hint.config
+---@field format_label? fun(label: string): string Formatting function for customizing an inlay hint's label
```
Comment (wookayin) [inline runtime/lua/vim/lsp/inlay_hint.lua:8]: What else context information (e.g. bufstate, ctx, or even the bufnr itself) could we add here?

File: runtime/lua/vim/lsp/inlay_hint.lua [hunk 2]
```diff
@@ -332,10 +341,16 @@ api.nvim_set_decoration_provider(namespace, {
             if type(label) == 'string' then
               text = label
             else
+              -- The label is made of diplay parts, so concatenate their values.
```
Comment (wookayin) [inline runtime/lua/vim/lsp/inlay_hint.lua:344]:
```suggestion
              -- The label is made of diplay parts, so concatenate their values.
              --- @cast label lsp.InlayHintLabelPart[]
```

Comment (wookayin) [reply runtime/lua/vim/lsp/inlay_hint.lua:344]: In addition, I think exposing `lsp.InlayHintLabelPart[]` to API (rather than the "concatenated" labels) would be a more general solution. So the user-custom label formatter can directly access `label: string | InlayHintLabelPart[]` as specified in [the LSP Protocol](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHint).

File: runtime/lua/vim/lsp/inlay_hint.lua (hunk 1 above)
Comment (MariaSolOs) [reply runtime/lua/vim/lsp/inlay_hint.lua:8]: I think we shouldn't leak more state, since then that would tie us to keep that state around even if the implementation changes.

File: runtime/lua/vim/lsp/inlay_hint.lua (hunk 2 above)
Comment (MariaSolOs) [reply runtime/lua/vim/lsp/inlay_hint.lua:344]: That's a good idea. We can also leave it to the user how they want to deal with display parts.

Comment (MariaSolOs) [reply runtime/lua/vim/lsp/inlay_hint.lua]: I'm not fully typing label parts for now since I don't think we have the definitions for all subtypes yet.

File: runtime/lua/vim/lsp/inlay_hint.lua [hunk 3]
```diff
@@ -4,11 +4,17 @@ local ms = require('vim.lsp.protocol').Methods
 local api = vim.api
 local M = {}
 
+---@class vim.lsp.inlay_hint.config
+---@field format_label? fun(label: string|table): string Formatting function for customizing an inlay hint's label
+
 ---@class lsp.inlay_hint.bufstate
 ---@field version? integer
----@field client_hint? table<integer, table<integer, lsp.InlayHint[]>> client_id -> (lnum -> hints)
+---@field client_hints? table<integer, table<integer, lsp.InlayHint[]>> client_id -> (lnum -> hints)
```
Comment (justinmk) [inline runtime/lua/vim/lsp/inlay_hint.lua:12]: This is a good change, maybe do in a separate PR to unblock it.

**Comment by justinmk** [comment]
> I really don't want to make the same mistakes from `vim.lsp.util` of providing redundant helpers.
> 
> wdy all think about this? Does it make sense to provide commonly-used LSP handler options even if they can be easily (< 10 lines of code) overriden?

`on_func` https://github.com/neovim/neovim/pull/22598 generalizes handler wrapping. Although https://github.com/neovim/neovim/pull/22598#issuecomment-1463367606 notes some limitations, it seems like `on_func` would work for "truncation"?

```lua
vim.on_func(vim.lsp.handlers, 'textDocument/inlayHint', function(fn, args)
  vim.notify('before, yay')
  local rv = fn(unpack(args))
  rv.hint.label = truncate(rv.hint.label)
  return rv
end)
````

File: runtime/lua/vim/lsp/inlay_hint.lua [hunk 4]
```diff
@@ -330,12 +341,18 @@ api.nvim_set_decoration_provider(namespace, {
             local text = ''
             local label = hint.label
             if type(label) == 'string' then
-              text = label
+              text = formatter and formatter(label) or label
             else
-              for _, part in ipairs(label) do
-                text = text .. part.value
+              -- The label is made of diplay parts, (we concatenate their values by default).
```
Comment (justinmk) [inline runtime/lua/vim/lsp/inlay_hint.lua:346]:
```suggestion
              -- The label is made of display parts, (we concatenate their values by default).
```

File: runtime/lua/vim/lsp/inlay_hint.lua [hunk 5]
```diff
@@ -17,10 +23,12 @@ local augroup = api.nvim_create_augroup('vim_lsp_inlayhint', {})
 
 --- |lsp-handler| for the method `textDocument/inlayHint`
 --- Store hints for a specific buffer and client
+---@param err table?
 ---@param result lsp.InlayHint[]?
 ---@param ctx lsp.HandlerContext
+---@param config vim.lsp.inlay_hint.config?
 ---@private
-function M.on_inlayhint(err, result, ctx, _)
+function M.on_inlayhint(err, result, ctx, config)
```
Comment (justinmk) [inline runtime/lua/vim/lsp/inlay_hint.lua:31]: `on_inlayhint` is not mentioned in `:help lsp`, I wonder if we need a list of these handlers near the list at `:help lsp-method`. Otherwise they are much less discoverable.

File: runtime/lua/vim/lsp/inlay_hint.lua (hunk 3 above)
Comment (MariaSolOs) [reply runtime/lua/vim/lsp/inlay_hint.lua:12]: Here it is: https://github.com/neovim/neovim/pull/27250 :)

File: runtime/lua/vim/lsp/inlay_hint.lua (hunk 5 above)
Comment (MariaSolOs) [reply runtime/lua/vim/lsp/inlay_hint.lua:31]: Right, that's because the function is marked as private. I was also unsure of how to document this, since it's being used in places not processed by luadoc rn...

**Comment by MariaSolOs** [comment]
> > I really don't want to make the same mistakes from `vim.lsp.util` of providing redundant helpers.
> > wdy all think about this? Does it make sense to provide commonly-used LSP handler options even if they can be easily (< 10 lines of code) overriden?
> 
> #22598 generalizes handler wrapping by introducing `on_func` to monkeypatch any function in a "stackable" way. Although [#22598 (comment)](https://github.com/neovim/neovim/pull/22598#issuecomment-1463367606) notes some limitations, it seems like `on_func` would work for "truncation"?
> 
> ```lua
> vim.on_func(vim.lsp.handlers, 'textDocument/inlayHint', function(fn, args)
>   vim.notify('before, yay')
>   local rv = fn(unpack(args))
>   rv.hint.label = truncate(rv.hint.label)
>   return rv
> end)
> ```

#22598 does look like a much better (and global) solution. I'm okay with closing this PR if we think it's not a proper way to manipulate LSP responses.

**Comment by justinmk** [comment]
This doesn't look out of the question (a format callback is much better than a special case flag), but we are also looking for a more generalized solution. We have made compromises for urgent use-cases, not sure about the urgency here.

**Comment by MariaSolOs** [comment]
I'll close this for now, since I agree in this not being the best solution for LSP handler customization.
